main.go

```diff
@@ -123,6 +128,26 @@ func main() {
 
 	var err error
 
+	//nolint:reassign // We need to reassign ErrorHandlers to register our handler
+	utilruntime.ErrorHandlers = append(utilruntime.ErrorHandlers, func(err error) {
+		if errors.Is(err, x509.UnknownAuthorityError{}) || errors.Is(err, x509.CertificateInvalidError{}) {
+			lastBadCertificateLock.Lock()
+			defer lastBadCertificateLock.Unlock()
+
+			var unknownAuthorityError x509.UnknownAuthorityError
+			var certificateInvalidError x509.CertificateInvalidError
+			if errors.As(err, &unknownAuthorityError) && unknownAuthorityError.Cert != lastBadCertificate {
+				logger.Errorf(err, "Certificate error: %s", resource.ToJSON(err))
+				lastBadCertificate = unknownAuthorityError.Cert
+			}
+			if errors.As(err, &certificateInvalidError) && certificateInvalidError.Cert != lastBadCertificate {
+				logger.Errorf(err, "Certificate error: %s", resource.ToJSON(err))
+				lastBadCertificate = certificateInvalidError.Cert
+			}
+			// The generic handler has already logged the error, no need to repeat if we don't want extra detail
+		}
```

Comment:
I've found `errors.Is` to be problematic - it only works for error types that define an `Is` method which these 2 errors don't. So we need to use `errors.As`. Also we can simplify a bit by using an `atomic.Value` for `lastBadCertificate`:

```
var lastBadCertificate atomic.Value
...

//nolint:reassign // We need to reassign ErrorHandlers to register our handler
utilruntime.ErrorHandlers = append(utilruntime.ErrorHandlers, func(err error) {
    var unknownAuthorityError x509.UnknownAuthorityError
    if errors.As(err, &unknownAuthorityError) && unknownAuthorityError.Cert != lastBadCertificate.Load() {
        logger.Errorf(err, "Certificate error: %s", resource.ToJSON(err))
	lastBadCertificate.Store(unknownAuthorityError.Cert)
    }

    var certificateInvalidError x509.CertificateInvalidError
    if errors.As(err, &certificateInvalidError) && certificateInvalidError.Cert != lastBadCertificate.Load() {
        logger.Errorf(err, "Certificate error: %s", resource.ToJSON(err))
	lastBadCertificate.Store(certificateInvalidError.Cert)
    }
})

```

Reply:
Ah yes, they don’t implement `Is()`. Thanks for the `atomic.Value` tip, but using separate `Load()` and `Store()` doesn’t provide the same semantics as a mutex; however `Swap()` does.

Reply:
It doesn't - it depends if you want strong or eventual consistency. Either way would be fine here.

Reply:
I meant that in this scenario, using `Swap()` ensures that there isn’t a race between comparing and storing the value, in the same way as protecting the whole function with a mutex.